web-services/security/src/test/java/datawave/security/login/DatawavePrincipalLoginModuleTest.java

```diff
@@ -269,7 +271,179 @@ public void testGetRoleSetsLeavesRequiredRoles() throws Exception {
         
         verifyAll();
     }
-    
+
+    @Test(expected = CredentialException.class)
+    public void testProxiedEntitiesLoginNoRole() throws Exception {
+        // Call Chain is U -> S1 -> S2.  S2 will have no role.  This test case tests
+        // the case of no role for the terminal service.  This should fail with
+        // CredentialException thrown.
+        String issuerDN = DnUtils.normalizeDN(testServerCert.getIssuerDN().getName());
+        String serverDN = DnUtils.normalizeDN("CN=testServer.example.com, OU=iamnotaperson, OU=acme");
+        SubjectIssuerDNPair server1 = SubjectIssuerDNPair.of(serverDN, issuerDN);
+        String otherServerDN = DnUtils.normalizeDN("CN=otherServer.example.com, OU=iamnotaperson, OU=acme");
+        SubjectIssuerDNPair server2 = SubjectIssuerDNPair.of(otherServerDN, issuerDN);
+        String proxiedSubjects = "<" + userDN.subjectDN() + "><" + otherServerDN + ">";
+        String proxiedIssuers = "<" + userDN.issuerDN() + "><" + issuerDN + ">";
+        DatawaveCredential datawaveCredential = new DatawaveCredential(testServerCert, proxiedSubjects, proxiedIssuers);
+        callbackHandler.name = datawaveCredential.getUserName();
+        callbackHandler.credential = datawaveCredential;
+
+        List<String> s1Roles = Arrays.asList("AuthorizedServer");
+
+
+        DatawaveUser s1 = new DatawaveUser(server1, UserType.SERVER, null, s1Roles, null, System.currentTimeMillis());
+        DatawaveUser s2 = new DatawaveUser(server2, UserType.SERVER, null, null, null, System.currentTimeMillis());
+        DatawavePrincipal expected = new DatawavePrincipal(Lists.newArrayList(defaultPrincipal.getPrimaryUser(), s1, s2));
+
+        expect(securityDomain.getKeyStore()).andReturn(serverKeystore);
+        expect(securityDomain.getTrustStore()).andReturn(truststore);
+        expect(datawaveUserService.lookup(datawaveCredential.getEntities())).andReturn(expected.getProxiedUsers());
+
+        replayAll();
+        boolean success = datawaveLoginModule.login();
+        assertFalse("Login did not succeed.", success);
```

Comment:
If the login is expected to fail, but succeeds... isn't this a strange message to print?

Reply:
Will change.  Must have missed the error message when I updated the tests.